include/deal.II/fe/fe_p1nc.h

```diff
@@ -0,0 +1,311 @@
+ *  +--------0.5--------+
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ * 0.5                 0.0
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  +--------0.0--------+
+ *  @endverbatim
+ *
+ * <li> shape function $\phi_3$:
+ *  @verbatim
+ *  +--------0.5--------+
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ * 0.0                 0.5
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  +--------0.0--------+
+ *  @endverbatim
+ *
+ * </ul>
+ *
+ * Note that above shape functions are constructed on each cell, not on the reference cell only.
+ * @p get_linear_shape computes the coefficients for shape functions when @p fill_fe_values
+ * is called on each cell.
+ *
+ * The basis function associated with a node in global scope is defined by the composition of
+ * (local) basis functions associated with the node on each element.
+ * And DOF associated with the node represents the coefficient of the basis function associated with the node.
+ * In this frame, all DOFs are truly independent.
+ * When a problem with homogeneous Dirichlet boundary condition is considered, the total number of DOFs
+ * is equal to the number of interior nodes, as the number of DOFs which the standard bilinear
+ * finite element @p Q_1 has.
+ *
+ * <h3>Unit support points</h3>
+ * Contrast with ordinary Lagrange finite elements, DOF value with respect to the P1 nonconforming
+ * element at given node does not coincide with the function value at that node.
+ * For instance, the basis function associated with a node has value 0.75 at that node, not 1.0.
+ * Thus we need an interpolation operator which maps any smooth function into a function
+ * with proper DOF values in the P1 element space.
+ * One natural interpolant associated with given smooth function is the linear function whose midpoint
+ * value at each edge is defined by the average of two values at endpoints of the edge.
+ * It provides appropriate weights used in @p unit_support_points.
+ *
+ * <h3>References</h3>
+ * You can find the paper about the P1NC element at
+ * http://epubs.siam.org/doi/abs/10.1137/S0036142902404923.
+ *
+ **/
+
+class FE_P1NC : public FiniteElement<2,2>
+{
+
+public:
+  /**
+   * Constructor for nonparametric version of P1 nonconforming element.
+   */
+  FE_P1NC() ;
+
+  /**
+   * Return the name of the class for the element.
+   */
+  virtual std::string get_name () const ;
+
+  /**
+   * Return the update flags which are needed.
+   */
+  virtual UpdateFlags requires_update_flags (const UpdateFlags flags) const ;
+
+  /**
+   * Copy constructor.
+   */
+  virtual FiniteElement<2,2> *clone () const ;
+
+  /**
+   * Destructor.
+   */
+  virtual ~FE_P1NC ();
+
+
+
+private:
+
+  static std::vector<ComponentMask> get_nonzero_component();
```

Comment:
I added this function because i follow another code for a staggered DG element implementation when i began. 
It is about component mask for advanced vector variables (right?) 
and I guess it is a convention for initialization of the class FiniteElement.